src/secp256k1.c

```diff
@@ -681,10 +730,144 @@ int secp256k1_ec_pubkey_combine(const secp256k1_context* ctx, secp256k1_pubkey *
     return 1;
 }
 
+/* Converts the point encoded by a secp256k1_pubkey into its absolute value,
+ * i.e. keeps it as is if it is positive and otherwise negates it. Sign is set
+ * to 1 if the input point was negative and set to 0 otherwise. */
+static void secp256k1_ec_pubkey_absolute(const secp256k1_context* ctx, secp256k1_pubkey *pubkey, int *sign) {
+   secp256k1_ge ge;
+    secp256k1_pubkey_load(ctx, &ge, pubkey);
+    if (sign != NULL) {
+        *sign = 0;
+    }
+    if (!secp256k1_fe_is_quad_var(&ge.y)) {
+        secp256k1_ge_neg(&ge, &ge);
+        if (sign != NULL) {
+            *sign = 1;
+        }
+    }
+    secp256k1_pubkey_save(pubkey, &ge);
+}
+
+int secp256k1_xonly_pubkey_create(const secp256k1_context* ctx, secp256k1_xonly_pubkey *pubkey, const unsigned char *seckey) {
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(seckey != NULL);
+
+    if (!secp256k1_ec_pubkey_create(ctx, (secp256k1_pubkey *) pubkey, seckey)) {
+        return 0;
+    }
+    secp256k1_ec_pubkey_absolute(ctx, (secp256k1_pubkey *) pubkey, NULL);
+    return 1;
+}
+
+int secp256k1_xonly_pubkey_parse(const secp256k1_context* ctx, secp256k1_xonly_pubkey* pubkey, const unsigned char *input32) {
+    /* TODO parse directly from 32 byte buffer */
+    unsigned char buf[33];
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(input32 != NULL);
+
+    buf[0] = SECP256K1_TAG_PUBKEY_EVEN;
+    memcpy(&buf[1], input32, 32);
+    if (!secp256k1_ec_pubkey_parse(ctx, (secp256k1_pubkey *) pubkey, buf, sizeof(buf))) {
+        return 0;
+    }
+    secp256k1_ec_pubkey_absolute(ctx, (secp256k1_pubkey *) pubkey, NULL);
+    return 1;
+}
+
+int secp256k1_xonly_pubkey_serialize(const secp256k1_context* ctx, unsigned char *output32, const secp256k1_xonly_pubkey* pubkey) {
+    /* TODO serialize directly into 32 byte buffer */
+    unsigned char buf[33];
+    size_t outputlen = sizeof(buf);
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(output32 != NULL);
+
+    if (!secp256k1_ec_pubkey_serialize(ctx, buf, &outputlen, (secp256k1_pubkey *) pubkey, SECP256K1_EC_COMPRESSED)) {
```

Comment:
Should be `(const secp256k1_pubkey *) pubkey` to avoid casting away the `const` qualifier.

Reply:
fixed